synnax/pkg/distribution/channel/channel_group.go

```diff
@@ -0,0 +1,273 @@
+//
+// channelGroup implements tools for constructing a set of channels. The set of
+// channels is constructed as a sorted array of channelRanges. Each channelRange
+// consists of the local key that the range starts at and the size of the
+// channelRange (how many consecutive keys it is).
+
+package channel
+
+import (
+	"slices"
+	"sort"
+)
+
+// channelRange is the basic struct for ChannelGroup, and is defined as a set of
+// consecutive channel keys.
+type channelRange struct {
+	// start is the local key that the channelRange starts at.
+	start uint16
+	// size is how many consecutive keys there are in a row.
+	size uint16
+}
+
+// end returns the key that is at the end of channelRange r
+func (r channelRange) end() uint16 {
+	return r.start + r.size - 1
+}
+
+// ChannelGroup extends []channelRange with a few convenience methods.
+type ChannelGroup []channelRange
+
+// Len returns the length of c. Implements sort.interface.
+func (c *ChannelGroup) Len() int {
+	return len(*c)
+}
+
+// Swap switches the places of (*c)[i] and (*c)[j]. Implements sort.interface.
+func (c *ChannelGroup) Swap(i, j int) {
+	temp := (*c)[i]
+	(*c)[i] = (*c)[j]
+	(*c)[j] = temp
+}
+
+// Less returns true if the key of (*c)[i] is before the key of (*c)[j].
+// Implements sort.interface.
+func (c *ChannelGroup) Less(i, j int) bool {
+	return (*c)[i].start < (*c)[j].start
+}
+
+// InsertKey allows you to insert one single key into the channelGroup c.
+func (c *ChannelGroup) InsertKey(key Key) {
+	c.InsertKeys(Keys{key})
+}
+
+// InsertKeys allows you to insert several keys into the channelGroup c.
+func (c *ChannelGroup) InsertKeys(keys Keys) {
+	if len(keys) == 0 {
+		return
+	}
+
+	localKeys := make([]uint16, len(keys))
+	for i, key := range keys {
+		localKeys[i] = key.LocalKey()
+	}
+
+	r := channelRange{
+		start: localKeys[0],
+		size:  0,
+	}
+	for i, key := range localKeys {
+		if r.start+r.size == key {
+			r.size++
+		} else {
+			c.insert(r)
+			r.start = localKeys[i]
+			r.size = 1
+		}
+	}
+	c.insert(r)
+}
+
+// RemoveKey removes key from c.
+func (c *ChannelGroup) RemoveKey(key Key) {
+	c.RemoveKeys(Keys{key})
+}
+
+// RemoveKeys removes all keys in keys from c.
+func (c *ChannelGroup) RemoveKeys(keys Keys) {
+	if len(keys) == 0 {
+		return
+	}
+
+	localKeys := make([]uint16, len(keys))
+	i := 0
+	for _, key := range keys {
+		if c.IsKeyIn(key) {
+			localKeys[i] = key.LocalKey()
+			i++
+		}
+	}
+	localKeys = localKeys[0:i]
+	if len(localKeys) == 0 {
+		return
+	}
+
+	r := channelRange{
+		start: localKeys[0],
+		size:  0,
+	}
+	for i, key := range localKeys {
+		if r.start+r.size == key {
+			r.size++
+		} else {
+			c.remove(r)
+			r.start = localKeys[i]
+			r.size = 1
+		}
+	}
+	c.remove(r)
+}
+
+// GetChannelsBeforeKey returns the number of channels in c before key. If key
+// is in c, that does not count towards the number of channels returned.
+func (c *ChannelGroup) GetChannelsBeforeKey(key Key) uint16 {

```

Comment:
This would be better named `GetNumChannelsBeforeKey` or `GetNumBeforeKey`. Remember when you're writing go code, you should ready all package internal methods in the context of their package name. If the method is called `GetNumBeforeKey` and the package name is `channel`, you should assume that it is the number of channels before a particular key.